### DISCUSSION TITLE
Location as a trigger? Maybe I'm missing something...

### DISCUSSION THREAD

**Discussion by michaelherger** [rxhanson/RectanglePro-Community#349] [General]
I'm taking my MacBook to various places with three pretty different setups:

* 4k monitor, MB closed
* MB plus two external monitors
* MB plus single ultra-wide monitor

macOS as we know doesn't remember positions well. When I walk away to grab a snack, I might return and find all applications crammed on the notebook display. That's where I hoped Rectangle Pro would help. First tests are promising. But sometimes the "wake up" trigger doesn't seem to be working. Maybe a timing issue? Monitors aren't known to the OS yet when the action is triggered?

The other situation I don't understand yet: if all my three layouts use the wake up trigger, that wouldn't work, would it? Does Rectangle Pro take into account the environment? Does it recognise there's no external monitor, therefore use the single screen layout etc.?

**Comment by michaelherger** [comment]
I've found the link to the undocumented command line features documentation 😀. Would there be a command to configure the trigger events for app layouts? That would allow me to configure them based on the environment without clicking around...

**Comment by rxhanson** [comment]
> Maybe a timing issue? Monitors aren't known to the OS yet when the action is triggered?

Yes, this is pretty challenging to get the timing right since it's dealing with display hardware timing. I'll have to revisit this and see if there are improvements I can make.

> The other situation I don't understand yet: if all my three layouts use the wake up trigger, that wouldn't work, would it? Does Rectangle Pro take into account the environment? Does it recognise there's no external monitor, therefore use the single screen layout etc.?

Displays are taken into account by your configuration. For example, if you have a layout with absolute positioning of a window on a certain display and that display is unavailable then it won't position that window. If you have multiple layouts triggered on the same display at the same time, then it would go through and apply both of them and only the last one applied would be where the window would end up (there's room for user error there). 

> I've found the link to the undocumented command line features documentation

I think you're talking about the URLs for executing an app layout on the [readme](https://github.com/rxhanson/RectanglePro-Community)? A lot of users make their own triggers using other apps that just call the URL for their app layout, if you're interested in going that route.

**Reply by michaelherger** [reply]
I've seen those commands, and they might help indeed. But I'd prefer to be able to use the "on wake" trigger. But I'd only want to have one of those configured per environment. Therefore I was hoping to be able to configure the events using an external script, to avoid to have to re-configure triggers myself. I'm often working in two of the three environments the same day.

**Reply by michaelherger** [reply]
Is there a way to script configuration changes? Not changes to the window layout, but some of the settings that can be found in the preferences, like eg. the trigger to be used for a given app layout?

**Reply by rxhanson** [reply]
There's nothing built into the app for it, but you could try writing a script for changing the defaults plist and automatically restarting the app. The plist is ~/Library/Preferences/com.knollsoft.Hookshot.plist, or you could use the `defaults` terminal command. Disclaimer: this route is not something I directly support either, and it could lead to frustration.

**Reply by michaelherger** [reply]
😁 - I am a glutton for punishment when it comes to abusing applications in ways I shouldn't... So thanks a lot for the hint! The plist's path is what I hadn't figured out.

**Comment by michaelherger** [comment]
FWIW: I'm about to move away to discover environments by network. This is working poorly when I eg. unplug the monitor to work on the kitchen table instead. My script now uses `system_profiler -json SPDisplaysDataType` to get display information instead. That makes much more sense. I then configure the triggers based on what monitors are connected.

**Reply by rxhanson** [reply]
Interesting. Thanks for following up, I'll give this avenue some thought when I'm back in this part of the code.
